enhancements/network/static-from-dhcp.md

```diff
@@ -0,0 +1,152 @@
+
+1. Deployers on baremetal don't want an outage of their DHCP server to cascade
+   into an outage of their OpenShift cluster.
+
+2. Edge deployments may not have access to a DHCP server due to the limited
+   capacity in those environments.
+
+### Goals
+
+A DHCP server will initially be required for deployment, but after that the
+cluster should function with or without the DHCP server.
+
+### Non-Goals
+
+Fully static deployments with no DHCP at any point. That will require
+significantly larger architectural changes.
+
+## Proposal
+
+When we receive a DHCP lease with an infinite expiration time, create a static
+configuration on the node that reflects the values provided by the DHCP server.
+
+### User Stories [optional]
+
+#### Story 1
+
+As the administrator of a baremetal cluster, I do not want a DHCP server outage
+to cascade into an outage of any other services.
+
+#### Story 2
+
+As the deployer of an Edge environment, my nodes may have limited or no
+connectivity back to a central DHCP server. This should not affect their
+functionality.
+
+### Implementation Details/Notes/Constraints [optional]
+
+We will add a NetworkManager dispatcher script that looks at each DHCP lease.
+If the lease duration is infinite, the script will take the values provided
+through DHCP and configure them statically on the node using nmcli. This
+means the interface will not be configured via DHCP from then on, but since
+we don't allow changing IPs on nodes anyway that shouldn't be a problem.
+
+### Risks and Mitigations
+
+If an address that was statically assigned in this way is reassigned by the
+DHCP server, it could cause a conflict. There are two mitigations to this:
+First, the requirement for the lease time to be infinite makes it unlikely
+the server would reassign it in the course of normal operation, and second,
+some DHCP servers have conflict detection that will prevent them from
+reassigning an address that is active on the network.
+
+Static configuration of the node's networking may make future changes more
+difficult. In general we don't support changing the IP address of a node
+so that isn't a concern at this time, but if routes or DNS search domains
+are provided by DHCP it won't be possible to update those either after
+initial deployment. Making such changes after deployment would require use
+of a different mechanism.
+
+There may be deployers who do not want their DHCP addresses configured
+statically, but do want to use infinite DHCP leases. This could be worked
+around by using extremely long leases that are not technically infinite,
+but in practice would never expire.
+
+## Design Details
+
+### Open Questions
+
+### Test Plan
+
+This should be testable in virt environments. The one concern would be
+whether our dev environment configurations can be configured to use
+infinite leases. However, I believe our test team has a way to test with
+external DHCP, so that should allow us to do what we need.
```

Comment:
I think we need to figure out where CI coverage for this can exist - maybe a new e2e-metal-ipi periodic job, or we could consider using this behavior by default in one of the existing jobs?

Reply:
I have a patch up to enable support for this in dev-scripts: https://github.com/openshift-metal3/dev-scripts/pull/1176

One option to test this in the existing jobs would be to set just one node to have an infinite lease. That way we'd have coverage of both paths. The only issue with that is our current version of libvirt chokes on infinite leases, which causes errors in some of the post-deploy steps. The cluster deploys fine, but as Derek noted on the dev-scripts pr it prevents things like clouds.yaml from being generated.

Reply:
Okay, I have recently discovered that the centos 8 advanced-virtualization repo now has a new enough version of libvirt to test this properly without hacking the libvirt dnsmasq config. I've updated the enhancement to suggest a periodic job since testing this will require things like rebooting the nodes without a DHCP server, which doesn't make sense for most changes we test.